<?xml version="1.0"?> 
<robot name="ram" xmlns:xacro="http://www.ros.org/wiki/xacro">

<!-- ...................Material Definitions........................ -->

  <material name="blue">
    <color rgba="0.0 0.0 0.8 1.0"/>
  </material>

  <material name="black">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>

<!--.................... Constant Definitions........................-->
  <xacro:property name="deg" value="0.017453293"/>

<!--.................... Fixing the robot to Gazebo.................. -->
  <link name="world"/>

  <joint name="fixed" type="fixed">
    <parent link="world"/>
    <child link="base_link"/>
  </joint>

<!---............................. Link Definitions ..................-->
<link name="base_link">
    <inertial>
	<origin xyz="0 0 0" rpy="0 0 0"/>
	<mass value="0.203873598369"/>
	<inertia ixx="0.00530920616718" ixy="0.0" ixz="0.0" iyy="0.00530920616718" iyz="0.0" izz="0.00849472986748"/>
    </inertial>

    <visual>
        <geometry>
           <box size="0.5 0.5 0.25"/>
        </geometry>
        <material name ="blue"/>
        <origin rpy="0 0 0" xyz="0 0 0"/>
    </visual>

    <collision>
	<origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
           <box size="0.5 0.5 0.25"/>
        </geometry>
    </collision>
</link>

<link name="link1">
    <inertial>
	<origin xyz="0 0 0.25" rpy="0 0 0"/>
	<mass value="0.101936799185"/>
	<inertia ixx="0.0010193675841" ixy="0.0" ixz="0.0" iyy="0.0010193675841" iyz="0.0" izz="0.000509683995923"/>
    </inertial>

    <visual>
        <geometry>
           <cylinder radius="0.1" length="0.3"/>
        </geometry>
        <material name ="black"/>
        <origin rpy="0 0 0" xyz="0 0 0.25"/>
    </visual>

    <collision>
	<origin xyz="0 0 0.25" rpy="0 0 0"/>
        <geometry>
           <cylinder radius="0.1" length="0.3"/>
        </geometry>
    </collision>
</link>

<link name="link2">
    <inertial>
	<origin xyz="0 0 0" rpy="${90*deg} 0 0"/>
        <mass value="0.0886850152905"/>
	<inertia ixx="0.000720565461009" ixy="0.0" ixz="0.0" iyy="0.000720565461009" iyz="0.0" izz="0.000110856269113"/>
    </inertial>

    <visual>
        <geometry>
	   <cylinder length="0.3" radius="0.05"/>
	</geometry>
        <material name ="black"/>
        <origin rpy="${90*deg} 0 0" xyz="0 0 0"/>
    </visual>

    <collision>
	<origin xyz="0 0 0" rpy="${90*deg} 0 0"/>
        <geometry>
	   <cylinder length="0.3" radius="0.05"/>
	</geometry>
    </collision>

</link>

<link name="link3">
    <inertial>
	<origin xyz="0 0 0" rpy="0 0 0"/>
	<mass value="0.0886850152905"/>
	<inertia ixx="0.0116629986477" ixy="0.0" ixz="0.0" iyy="0.0116629986477" iyz="0.0" izz="0.000230950468272"/>
    </inertial>

    <visual>
        <geometry>
           <box size="0.125 0.125 1.25"/>
        </geometry>
        <material name ="blue"/>
        <origin rpy="0 0 0" xyz="0 0 0"/>
    </visual>

    <collision>
	<origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
           <box size="0.125 0.125 1.25"/>
        </geometry>
    </collision>

</link>

<link name="link4">
    <inertial>
	<origin xyz="0 0 0" rpy="${90*deg} 0 0"/>
	<mass value="0.0662589194699"/>
	<inertia ixx="0.000538353505352" ixy="0.0" ixz="0.0" iyy="0.000538353505352" iyz="0.0" izz="8.28236493374e-05"/>
    </inertial>

    <visual>
        <geometry>
	   <cylinder length="0.3" radius="0.05"/>
	</geometry>
        <material name ="black"/>
        <origin rpy="${90*deg} 0 0" xyz="0 0 0"/>
    </visual>

    <collision>
	<origin xyz="0 0 0" rpy="${90*deg} 0 0"/>
        <geometry>
	   <cylinder length="0.3" radius="0.05"/>
	</geometry>
    </collision>

</link>

<link name="link5">
    <inertial>
	<origin xyz="0 0 0" rpy="0 0 0"/>
	<mass value="0.0540265035678"/>
	<inertia ixx="0.00457255381148" ixy="0.0" ixz="0.0" iyy="0.00457255381148" iyz="0.0" izz="0.00014069396343"/>
    </inertial>

    <visual>
        <geometry>
           <box size="0.125 0.125 1"/>
        </geometry>
        <material name ="blue"/>
        <origin rpy="0 0 0" xyz="0 0 0"/>
    </visual>

    <collision>
	<origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
           <box size="0.125 0.125 1"/>
        </geometry>
    </collision>

</link>

<link name="link6">
    <inertial>
	<origin xyz="0 0 0" rpy="${90*deg} 0 0"/>
	<mass value="0.0489296636086"/>
	<inertia ixx="9.17430825688e-05" ixy="0.0" ixz="0.0" iyy="9.17430825688e-05" iyz="0.0" izz="0.000137614678899"/>
    </inertial>

    <visual>
        <geometry>
           <cylinder radius="0.075" length="0.075"/>
        </geometry>
        <material name ="black"/>
        <origin rpy="${90*deg} 0 0" xyz="0 0 0"/>
    </visual>

    <collision>
	<origin xyz="0 0 0" rpy="${90*deg} 0 0"/>
        <geometry>
           <cylinder radius="0.075" length="0.075"/>
        </geometry>
    </collision>

</link>

<link name="link7">
    <inertial>
	<origin xyz="0 0 0" rpy="0 0 0"/>
	<mass value="0.032619775739"/>
	<inertia ixx="0.00117567061366" ixy="0.0" ixz="0.0" iyy="0.00117567061366" iyz="0.0" izz="5.43662711519e-05"/>
    </inertial>

    <visual>
        <geometry>
           <box size="0.1 0.1 0.65"/>
        </geometry>
        <material name ="blue"/>
        <origin rpy="0 0 0" xyz="0 0 0"/>
    </visual>

    <collision>
	<origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
           <box size="0.1 0.1 0.65"/>
        </geometry>
    </collision>

</link>

<!--...........................Joint Definitions.....................................-->
    <joint name="joint1" type="revolute">
        <parent link="base_link"/>
        <child link="link1"/>
        <axis xyz="0 0 1" />
        <limit effort="420.0" lower="${-180*deg}" upper="${180*deg}" velocity="${30*deg}" />
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <dynamics damping="0.7"/>
    </joint>

    <joint name="joint2" type="revolute">
        <parent link="link1"/>
        <child link="link2"/>
        <axis xyz="0 1 0" />
        <limit effort="150.0" lower="${-90*deg}" upper="${90*deg}" velocity="${15*deg}" />
        <origin xyz="0 0.15 0.3" rpy="0 0 0"/>
        <dynamics damping="0.7"/>
    </joint>

    <joint name="joint3" type="revolute">
	<parent link="link2"/>
	<child link="link3"/>
        <limit effort="0" lower="0" upper="0" velocity="0"/>
        <origin xyz="0 0.05 0.5" rpy="0 0 0"/>
    </joint>

    <joint name="joint4" type="revolute">
        <parent link="link3"/>
        <child link="link4"/>
        <axis xyz="0 1 0" />
        <limit effort="100.0" lower="${-135*deg}" upper="${135*deg}" velocity="${23*deg}" />
        <origin xyz="0 -0.1 0.5" rpy="0 0 0"/>
        <dynamics damping="0.7"/>
    </joint>

    <joint name="joint5" type="revolute">
	<parent link="link4"/>
	<child link="link5"/>
        <limit effort="0" lower="0" upper="0" velocity="0"/>
        <origin xyz="0 -0.08 0.4" rpy="0 0 0"/>
    </joint>

    <joint name="joint6" type="revolute">
        <parent link="link5"/>
        <child link="link6"/>
        <axis xyz="0 1 0" />
        <limit effort="100.0" lower="${-180*deg}" upper="${180*deg}" velocity="${30*deg}" />
        <origin xyz="0 0.08 0.4" rpy="0 0 0"/>
        <dynamics damping="0.7"/>
    </joint>

    <joint name="joint7" type="prismatic">
        <parent link="link6"/>
	<child link="link7"/>
        <axis xyz="0 0 1" />
        <limit effort="100" lower="0" upper="0.5" velocity="0.05"/>
        <origin xyz="0 0.08 -0.25" rpy="0 0 0"/>
        <dynamics damping="0.7"/>
    </joint>

<!--........................... Transmission Definitions.....................................-->
  <transmission name="trans1">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint1">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor1">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="trans2">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint2">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor2">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="trans3">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint3">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor3">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="trans4">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint4">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor4">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="trans5">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint5">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor5">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="trans6">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint6">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor6">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="trans7">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint7">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor7">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

<!--.................. Gazebo links............................. -->
  <gazebo reference="base_link">
    <material>Gazebo/Blue</material>
    <mu1>2</mu1>
    <mu2>2</mu2>
  </gazebo>

  <gazebo reference="link1">
    <material>Gazebo/Black</material>
    <mu1>2</mu1>
    <mu2>2</mu2>
  </gazebo>

  <gazebo reference="link2">
    <material>Gazebo/Black</material>
    <mu1>2</mu1>
    <mu2>2</mu2>
  </gazebo>

  <gazebo reference="link3">
    <material>Gazebo/Blue</material>
    <mu1>2</mu1>
    <mu2>2</mu2>
  </gazebo>

  <gazebo reference="link4">
    <material>Gazebo/Black</material>
    <mu1>2</mu1>
    <mu2>2</mu2>
  </gazebo>

  <gazebo reference="link5">
    <material>Gazebo/Blue</material>
    <mu1>2</mu1>
    <mu2>2</mu2>
  </gazebo>

  <gazebo reference="link6">
    <material>Gazebo/Black</material>
    <mu1>2</mu1>
    <mu2>2</mu2>
  </gazebo>

  <gazebo reference="link7">
    <material>Gazebo/Blue</material>
    <mu1>2</mu1>
    <mu2>2</mu2>
  </gazebo>

<!--................. Gazebo plugins....................-->
  <gazebo>
	<plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
	    <robotNamespace>/ram</robotNamespace>
	</plugin>
  </gazebo>

</robot>
